src/Zynga/Framework/IO/Disk/V1/Manager.hh

```
@@ -0,0 +1,243 @@
+   */
+  public function bzip2(string $in, string $out): void {
+    if (!$this->doesFileExist($in) || !$this->isReadable($in)) {
+      throw new ReadPermissionsException($in);
+    }
+
+    if ((!$this->doesFileExist($out) && !$this->isWriteable($this->directoryName($out)) ||
+        ($this->doesFileExist($out) && !$this->isWriteable($out)))) {
+      throw new WritePermissionsException($out);
+    }
+
+    $inFile = $this->fileOpen($in, "r");
+    if ($inFile === false) {
+      throw new FailedToOpenFileException($in);
+    }
+
+    $outFile = $this->bzopen($out, "w");
+    if ($outFile === false) {
+      $this->fclose($inFile);
+      throw new FailedToOpenFileException($out);
+    }
+
+    while (!$this->feof($inFile)) {
+      $buffer = $this->fgets($inFile, 4096);
+      $this->bzwrite($outFile, $buffer, 4096);
+    }
+
+    if (!$this->fclose($inFile)) {
+      $this->bzclose($outFile);
+      throw new FailedToCloseFileException($in);
+    }
+
+    if (!$this->bzclose($outFile)) {
+      throw new FailedToCloseFileException($out);
+    }
+  }
+
+  // Have to use mixed for $handle here because Hack doesn't allow
+  // passing resource types to functions
+  protected function feof(mixed $handle): bool {
+    if (is_resource($handle)) {
+      return feof($handle);
+    }
+
+    return false;
+  }
+
+  // Have to use mixed for $handle here because Hack doesn't allow
+  // passing resource types to functions
+  protected function fgets(mixed $handle, int $maxBytesToRead): string {
+    if (is_resource($handle)) {
+      $result = fgets($handle, $maxBytesToRead);
+      if (is_string($result)) {
+        return $result;
+      }
+    }
+
+    return '';
+  }
+
+  protected function doesFileExist(string $fileName): bool {
+    return file_exists($fileName);
+  }
+
+  protected function fileOpen(string $fileName, string $mode): mixed {
+    try {
+      return fopen($fileName, $mode);
+    } catch (Exception $e) {
+      return false;
```

Why not just rethrow the exception using your wrapped domain one?

```
 try {
      return fopen($fileName, $mode);
    } catch (Exception $e) {
      throw new FailedToOpenFileException("Failed to open file=$fileName", 0, $e);
}